rust/kernel/device.rs

```diff
@@ -74,3 +84,122 @@ impl Drop for Device {
         unsafe { bindings::put_device(self.ptr) };
     }
 }
+
+/// Device data.
+///
+/// When a device is removed (for whatever reason, for example, because the device was unplugged or
+/// because the user decided to remove the driver), the driver is given a chance to clean its state
+/// up, and all io resources should ideally not be used anymore.
+///
+/// However, the device data is reference-counted because other subsystems hold pointers to it. So
+/// some device state must be freed and not used anymore, while others must remain accessible.
+///
+/// This struct separates the device data into three categories:
+/// 1. Io resources: are available until the device is removed.
+/// 2. Registrations: are destroyed on when the device is removed, but before the io resources
+///    become inaccessible.
+/// 3. General data: remain available as long as the ref count is nonzero.
+///
+/// This struct implements the `DeviceRemoval` trait so that it can clean resources up even if not
+/// explicitly called by the device drivers.
+pub struct Data<Reg, Res, Gen> {
+    regs: RevocableMutex<Reg>,
+    res: Revocable<Res>,
+    general: Gen,
+}
+
+/// Safely creates an new reference-counted instance of [`Data`].
+#[doc(hidden)]
+#[macro_export]
+macro_rules! new_device_data {
+    ($reg:expr, $res:expr, $gen:expr, $name:literal) => {{
```

Comment:
I think there is the potential for people to mess up the order of the fields. Maybe do

```rust
let data = new_device_data! {
    registration: reg,
    resource: res,
    general: gen,
    name: "foo",
};
```

? With a potential shorthand when the variable name matches the field name just like with struct.

Reply:
The types are different and need to be spelled out elsewhere (as opposed to being inferred) because this is usually the `Data` type of a `Driver` trait. Here's an example of it being used: https://github.com/Rust-for-Linux/linux/pull/571/files#diff-0b4ec144d8f0479a497eca5a2e4ab8c1009b0debcdb00ece09f79e5baa8faac9R272

IOW, if people mess up the order, they'll get a compilation error.